src/libraries/System.IO/tests/BinaryWriter/BinaryWriter.EncodingTests.cs

```diff
@@ -0,0 +1,236 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Numerics;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Text;
+using Moq;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public class BinaryWriter_EncodingTests
+    {
+        [Fact]
+        public void Ctor_Default_UsesFastUtf8()
+        {
+            BinaryWriter writer = new BinaryWriter(new MemoryStream());
+            Assert.True(IsUsingFastUtf8(writer));
+        }
+
+        [Fact]
+        public void Ctor_EncodingUtf8Singleton_UsesFastUtf8()
+        {
+            BinaryWriter writer = new BinaryWriter(new MemoryStream(), Encoding.UTF8);
+            Assert.True(IsUsingFastUtf8(writer));
+        }
+
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void Ctor_NewUtf8Encoding_UsesFastUtf8(bool emitIdentifier, bool throwOnInvalidBytes)
+        {
+            BinaryWriter writer = new BinaryWriter(new MemoryStream(), new UTF8Encoding(emitIdentifier, throwOnInvalidBytes));
+            Assert.True(IsUsingFastUtf8(writer));
+        }
+
+        [Fact]
+        public void Ctor_Utf8EncodingWithSingleCharReplacementChar_UsesFastUtf8()
+        {
+            Encoding encoding = Encoding.GetEncoding("utf-8", new EncoderReplacementFallback("x"), DecoderFallback.ExceptionFallback);
+            BinaryWriter writer = new BinaryWriter(new MemoryStream(), encoding);
+            Assert.True(IsUsingFastUtf8(writer));
+        }
+
+        [Fact]
+        public void Ctor_Utf8EncodingWithMultiCharReplacementChar_DoesNotUseFastUtf8()
+        {
+            Encoding encoding = Encoding.GetEncoding("utf-8", new EncoderReplacementFallback("xx"), DecoderFallback.ExceptionFallback);
+            BinaryWriter writer = new BinaryWriter(new MemoryStream(), encoding);
+            Assert.False(IsUsingFastUtf8(writer));
+        }
+
+        [Fact]
+        public void Ctor_NotUtf8EncodingType_DoesNotUseFastUtf8()
+        {
+            BinaryWriter writer = new BinaryWriter(new MemoryStream(), new UnicodeEncoding());
+            Assert.False(IsUsingFastUtf8(writer));
+        }
+
+        [Fact]
+        public void Ctor_Utf8EncodingDerivedTypeWithWrongCodePage_DoesNotUseFastUtf8()
+        {
+            Mock<UTF8Encoding> mockEncoding = new Mock<UTF8Encoding>();
+            mockEncoding.Setup(o => o.CodePage).Returns(65000 /* UTF-7 code page */);
```

Comment:
I am ok with using `Mock<T>` as long as we don't have any AOT test suite that is going to fail.

![obraz](https://user-images.githubusercontent.com/6011991/105454625-79914700-5c82-11eb-967a-69e486595e52.png)

Reply:
I do not think it is worth it to pick up this heavy dependency here to just save like 3 lines.

For cases where it is really worth, we just need a way how to conditionally disable tests using test techniques that are incompatible with runtime mode (single file, trimming, no JIT, no reflection emit, no private reflection, etc.).

Reply:
We use Moq in a few other test projects in this repo (see [search results](https://github.com/dotnet/runtime/search?q=moq+extension%3Acsproj&type=code)). I can remove the dependency for this project, but what does that mean for the general test framework guidance?

See Steve's comment at https://github.com/dotnet/runtime/pull/47316#discussion_r562887714 and my response there for a little more context on why I'm using (and mocking) the `CodePage` property in the first place. We could tweak that logic and render the whole thing moot.

Reply:
I think using heavy test framework for testing high-level libraries like `Microsoft.Extensions.*` is fine. 

I do not think it is a good practice to use these heavy test frameworks for testing core platform (ie stuff in CoreLib).

I agree that we should have this test (as long as the implementation stays what it is). Do you agree that the use of Moq saves you like 3 lines on code in this case?